sw-emulator/lib/cpu/src/cpu.rs

```diff
@@ -15,11 +15,36 @@ Abstract:
 use crate::csr_file::{Csr, CsrFile};
 use crate::types::{RvInstr, RvMStatus};
 use crate::xreg_file::{XReg, XRegFile};
+use bit_vec::BitVec;
 use caliptra_emu_bus::{Bus, BusError, Clock, TimerAction};
 use caliptra_emu_types::{RvAddr, RvData, RvException, RvSize};
 
 pub type InstrTracer<'a> = dyn FnMut(u32, RvInstr) + 'a;
 
+#[derive(Clone)]
+pub struct CodeCoverage {
+    bit_vec: bit_vec::BitVec,
+}
+impl CodeCoverage {
+    pub fn new(capacity_in_bits: usize) -> Self {
+        Self {
+            bit_vec: BitVec::from_elem(capacity_in_bits, false),
+        }
+    }
+    pub fn log_execution(&mut self, pc: RvData) {
+        if (pc as usize) < self.bit_vec.len() {
+            self.bit_vec.set(pc as usize, true);
```

Comment:
Since instructions are 16-bit aligned, you could half the size of your bitvec and use (pc as usize / 2) as the index. Also when you get a 32-bit risc-v instruction, be sure to set all the bits associated with the instruciton.

Reply:
Done. Considering size of instruction and marking all bytes corresponding to a instruction as true.